_posts/2021-01-01-getting-started-with-graphql.md

```diff
@@ -0,0 +1,288 @@
+
+## Adding the Dependencies
+For GraphQL server, we will add the following Maven dependencies: 
+
+```xml
+    <dependency>
+      <groupId>com.graphql-java</groupId>
+      <artifactId>graphql-spring-boot-starter</artifactId>
+      <version>5.0.2</version>
+    </dependency>
+    <dependency>
+      <groupId>com.graphql-java</groupId>
+      <artifactId>graphql-java-tools</artifactId>
+      <version>5.2.4</version>
+    </dependency>
+```
+Here we have added `graphql-spring-boot-starter` as a GraphQL starter and a Java tools module `graphql-java-tools`.
+
+### Defining the GraphQL Schema 
+We can either take a top-down approach by defining the schema and then create POJO for each type or a bottom-up approach by creating the POJOs first followed by schema from the POJO classes. 
+
+The GraphQL schema needs to be defined in a file with the extension `graphqls`. Let us define our schema in a file `product.graphqls`:
+
+```
+type Product {
+    id: ID!
+    title: String!
+    description: String!
+    category: String
+    madeBy: Manufacturer!
+}
+
+type Manufacturer {
+    id: ID!
+    name: String!
+    address: String
+}
+
+# The Root Query for the application
+type Query {
+    myRecentPurchases(count: Int, customerID: String): [Product]!
+    lastVisitedProducts(count: Int, customerID: String): [Product]!
+    productsByCategory(category: String): [Product]!
+}
+
+# The Root Mutation for the application
+type Mutation {
+    addRecentProduct(title: String!, description: String!, category: String) : Product!
+}
+
+```
+Here we have added three operations to our Query and a Mutation for adding recent products.
+
+Next, we define the POJO classes for the Object types: `Product` and `Manufacturer`:
+
+```java
+@Data
+public class Product {
+   private String id; 
+   private String title;
+   private String description; 
+   private String category;
+   private Manufacturer madeBy;
+}
+```
+This `Product` POJO maps to the `product` type and `Manufacturer` maps to the `manufacturer` object defined in our GraphQL schema.
+
+
+### Associate GraphQL Types with Resolvers
+
+Multiple resolver components convert the GraphQl request received from the API consumers or front end and invoke operations to fetch data from applicable data sources. For each type, we define a `resolver`.
+
+We will now add resolvers for all the types defined in the schema. We first add a resolver class named `QueryResolver` containing the methods corresponding to our GraphQL operations: 
+
+```java
+@Service
+public class QueryResolver implements GraphQLQueryResolver {
```

Comment:
How is this `QueryResolver` mapped to the `Query` type in the schema? I guess this is Spring-magic? Can you explain this in a sentence? I.e. something like "if we put the schema file here, Spring will look for a `GraphQLQueryResolver` and use it to resolve any queries sent by the client".